src/rmp/src/Restructure.cpp

```diff
@@ -155,53 +160,138 @@ void Restructure::runABC()
+             1,
+             "Running abc with number of threads = {}",
+             max_threads);
+
+  for (int curr_mode_idx = 0; curr_mode_idx < modes.size();) {
+    int max_parallel_runs = (max_threads < modes.size() - curr_mode_idx)
+                                ? max_threads
+                                : modes.size() - curr_mode_idx;
+
+    // Spawn ABC process(es)
+    for (int curr_thread = 0; curr_thread < max_parallel_runs; ++curr_thread) {
+      int temp_mode_idx = curr_mode_idx + curr_thread;
+      output_blif_file_name_
+          = work_dir_name_ + std::string(block_->getConstName())
+            + std::to_string(temp_mode_idx) + "_crit_path_out.blif";
+
+      opt_mode_ = modes[temp_mode_idx];
+
+      std::string abc_script_file = work_dir_name_
+                                    + std::to_string(temp_mode_idx)
+                                    + "ord_abc_script.tcl";
+
+      debugPrint(logger_,
+                 RMP,
+                 "remap",
+                 1,
+                 "Writing ABC script file {}",
+                 abc_script_file);
+      if (writeAbcScript(abc_script_file)) {
+        std::string abc_command = std::string("yosys-abc < ") + abc_script_file;
+        if (logfile_ != "")
+          abc_command
+              = abc_command + " > " + logfile_ + std::to_string(temp_mode_idx);
+
+        pid_t child_pid = fork();
+        if (child_pid == 0) { // Begin child
+          // Run in child process
+          int ret = execlp("sh", "sh", "-c", abc_command.c_str(), 0);
+          // Execution of command failed
+          logger_->warn(RMP,
+                        31,
+                        "Failed to run ABC with exit code {}. Please check the "
+                        "messages for details.",
+                        ret);
+          exit(ret);
+        } // End child
+
+        if (child_pid > 0) {
+          child_proc[temp_mode_idx] = child_pid;
+        } else if (child_pid < 0) {
+          logger_->warn(RMP,
+                        29,
+                        "Failed to create new ABC process. Please check the "
+                        "messages for details.");
+        }
+
+        files_to_remove.emplace_back(abc_script_file);
+      }
+    }  // end spawn
+
+    // Wait for ABC process(es)
+    for (int curr_thread = 0; curr_thread < max_parallel_runs; ++curr_thread) {
+      int child_idx = curr_mode_idx + curr_thread;
+      pid_t child = child_proc[child_idx];
+
+      if (child == 0) {
+        continue;
+      }
+
+      int return_status;
+      waitpid(child, &return_status, 0);
+
+      if (return_status) {
+        child_proc[child_idx] = 0;
         logger_->warn(
             RMP,
-            11,
-            "ABC failed with code {}, please check messages for details.",
-            ret);
-        continue;
+            15,
+            "ABC ({}) failed with code {}, please check messages for details.",
+            child_idx,
+            return_status);
```

Comment:
Would an identifier inside the parentheses, e.g., `"ABC (process id = {})..."`, be helpful? Otherwise is just an integer out-of-context.

Also use a consistent format with period or comma before `please check...`.

Reply:
The process id wouldn't be helpful in this instance because after the execution is over for ABC, the process will just be killed and the pid wouldn't be of any use. The child_idx here has more meaning because it refers to the run id and its associated with a distinct logfile for each child_idx and that's how you use it.

Reply:
If the child index is printed the message is misleading because it is not an error code.
How is a user supposed to know where to "check messages"? That is completely vague.
I am going to guess that it isn't really messages that they should check, but a log file.

Reply:
I agree with check messages and will add the log file name that should be checked. For the child index, removing it would indicate that all abc runs have failed. I'm referring here to a single ABC run and the child idx is merely the id which is also associated with a particular log file for each one.